javascript_nodejs/14.nlp-with-dispatch/dialogs/qna/index.js

```diff
@@ -0,0 +1,41 @@
+const { QnAMaker } = require('botbuilder-ai');
+
+// this is the LUIS service type entry in the .bot file.
+const QnA_CONFIGURATION = 'sample.qna';
+const QnA_TOP_N = 1;
+const QnA_CONFIDENCE_THRESHOLD = 0.5;
+class QnADialog {
+    /**
+     * 
+     * @param {Object} botConfig bot configuration from .bot file
+     */
+    constructor(botConfig) {
+        if(!botConfig) throw ('Need bot config');
+
+        // add recogizers
+        const qnaConfig = botConfig.findServiceByNameOrId(QnA_CONFIGURATION);
+        if(!qnaConfig || !qnaConfig.kbId) throw (`QnA Maker application information not found in .bot file. Please ensure you have all required QnA Maker applications created and available in the .bot file. See readme.md for additional information\n`);
+        this.qnaRecognizer = new QnAMaker({
+            knowledgeBaseId: qnaConfig.kbId,
+            endpointKey: qnaConfig.endpointKey,
+            host: qnaConfig.hostname
+        });
```

Comment:
This pattern repeats in all samples and it boilerplate code. 
const qnaConfig = botConfig.findServiceByNameOrId(QnA_CONFIGURATION);
        if(!qnaConfig || !qnaConfig.kbId) throw (`QnA Maker application information not found in .bot file. Please ensure you have all required QnA Maker applications created and available in the .bot file. See readme.md for additional information\n`);
        this.qnaRecognizer = new QnAMaker({
            knowledgeBaseId: qnaConfig.kbId,
            endpointKey: qnaConfig.endpointKey,
            host: qnaConfig.hostname
        });

We should consider _initServiceByNameOrId_ for common services the Bot Framework supports like LUIS and QnA. Generic services can continue to use findServiceByNameOrId

this.qnaRecognizer = botConfig.initServiceByNameOrId(...) --> this code can throw on error

Reply:
A better option would be something like new LuisRecognizer(botConfig.findServiceByNameOrId(QNA_CONFIGURATION)) and have this throw instead. Filed microsoft/botbuilder-dotnet#908 to track this.